tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/MicroSecEndTest.java

```diff
@@ -0,0 +1,296 @@
+        flowGraph.evaluate();
+        
+        //Sanity check if cycles get detected
+        if(model.endsWith("sqshq_11")) {
+            assertTrue(flowGraph.wasCyclic());
+        }
+
+        assertFalse(flowGraph.getTransposeFlowGraphs()
+                .isEmpty());
+
+        Map<Integer, List<AbstractTransposeFlowGraph>> existenceViolations = new HashMap<>();
+        for (var transposeFlowGraph : flowGraph.getTransposeFlowGraphs()) {
+
+            hasLoggingServer(transposeFlowGraph, existenceViolations);
+
+            hasSecretManager(transposeFlowGraph, existenceViolations);
+
+            for (var vertex : transposeFlowGraph.getVertices()) {
+
+                hasGateway(violationsSet, variant, vertex);
+
+                hasAuthenticatedRerquest(violationsSet, variant, vertex);
+
+                hasAuthorizedEntrypoint(violationsSet, variant, vertex);
+
+                hasTransformedEntryIdentity(violationsSet, variant, vertex);
+
+                hasTokenValidation(violationsSet, variant, vertex);
+
+                hasLoginAttemptsRegulation(violationsSet, variant, vertex);
+
+                hasEncryptedEntryConnection(violationsSet, variant, vertex);
+
+                hasEncrytedInternalConnection(violationsSet, variant, vertex);
+
+                hasLocalLogging(violationsSet, variant, vertex);
+
+                hasLogSanitization(violationsSet, variant, vertex);
+
+                hasMessageBroker(violationsSet, variant, vertex);
+
+            }
+        }
+        var numOfTransposeFlowGraphs = flowGraph.getTransposeFlowGraphs()
+                .size();
+        if (existenceViolations.get(9)
+                .size() >= numOfTransposeFlowGraphs) {
+            violationsSet.add(9);
+            violationsSet.add(12);
+        }
+
+        if (existenceViolations.get(18)
+                .size() >= numOfTransposeFlowGraphs) {
+            violationsSet.add(18);
+        }
+    }
+
+    @Test
+    void testConstraints() {
+        for (var model : TUHH_MODELS.keySet()) {
+            for (int variant : TUHH_MODELS.get(model)) {
+                Set<Integer> violationSet = new TreeSet<Integer>();
+                String variationName = model + "_" + variant;
+                performAnalysis(Paths.get(location, model, variationName)
+                        .toString(), variant, violationSet);
+                logger.info("Variant: " + variationName);
+                logger.info("Violations: " + violationSet);
+                assertFalse(violationSet.contains(variant));
+            }
+        }
+    }
+
+    private boolean hasNodeWithCharacteristic(AbstractTransposeFlowGraph aTFG, String constraintRuleType, String constraintRule) {
+        return aTFG.stream()
+                .anyMatch(node -> hasNodeCharacteristic(node, constraintRuleType, constraintRule));
+    }
+
+    private boolean hasDataCharecteristicViolation(AbstractVertex<?> node, String constraintPrequisitType, String constraintPrequisit,
```

Comment:
If someone only reads the function name and sees the four strings, it's not quite clear what this function does.
It's essentially an implication check; if one Variable contains a label, it must contain the other label aswell.
Maybe we can find a better naming?

Reply:
I second that, I hade problems understand this method

Reply:
maybe dataCharacteristicImplicationCheck?

Reply:
how about checkLabelImplication? the verb should always be at the beginning.